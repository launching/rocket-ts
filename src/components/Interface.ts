export interface FormModel {
  [name: string]: any;
}

export interface Validate {
  [name: string]: any;
}
export interface FormItem {
  name: string;
  widget: string;
  validate: Array<any> | object;
}

export interface ButtonItem {
  text: string;
  type: string;
  action: string | Function;
  premise: Function;
}

export interface WidgetStore {
  [name: string]: any;
}

export interface Column {
  label: string;
  name: string;
  type: string;
  children: any;
}
