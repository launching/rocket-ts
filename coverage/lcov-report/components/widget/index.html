<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for components/widget</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> components/widget
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">73.45% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>83/113</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">16.33% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>8/49</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">48.15% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>13/27</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">69.7% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>69/99</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line medium'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="BaseWidget.ts"><a href="BaseWidget.ts.html">BaseWidget.ts</a></td>
	<td data-value="91.67" class="pic high"><div class="chart"><div class="cover-fill" style="width: 91%;"></div><div class="cover-empty" style="width:9%;"></div></div></td>
	<td data-value="91.67" class="pct high">91.67%</td>
	<td data-value="24" class="abs high">22/24</td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	<td data-value="90.48" class="pct high">90.48%</td>
	<td data-value="21" class="abs high">19/21</td>
	</tr>

<tr>
	<td class="file medium" data-value="Checkbox.vue"><a href="Checkbox.vue.html">Checkbox.vue</a></td>
	<td data-value="66.67" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 66%;"></div><div class="cover-empty" style="width:34%;"></div></div></td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="9" class="abs medium">6/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="3" class="abs low">1/3</td>
	<td data-value="62.5" class="pct medium">62.5%</td>
	<td data-value="8" class="abs medium">5/8</td>
	</tr>

<tr>
	<td class="file low" data-value="DatePickerGroup.vue"><a href="DatePickerGroup.vue.html">DatePickerGroup.vue</a></td>
	<td data-value="44.44" class="pic low"><div class="chart"><div class="cover-fill" style="width: 44%;"></div><div class="cover-empty" style="width:56%;"></div></div></td>
	<td data-value="44.44" class="pct low">44.44%</td>
	<td data-value="27" class="abs low">12/27</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="17" class="abs low">0/17</td>
	<td data-value="20" class="pct low">20%</td>
	<td data-value="5" class="abs low">1/5</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="25" class="abs low">10/25</td>
	</tr>

<tr>
	<td class="file high" data-value="Input.vue"><a href="Input.vue.html">Input.vue</a></td>
	<td data-value="90.91" class="pic high"><div class="chart"><div class="cover-fill" style="width: 90%;"></div><div class="cover-empty" style="width:10%;"></div></div></td>
	<td data-value="90.91" class="pct high">90.91%</td>
	<td data-value="11" class="abs high">10/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="88.89" class="pct high">88.89%</td>
	<td data-value="9" class="abs high">8/9</td>
	</tr>

<tr>
	<td class="file medium" data-value="Number.vue"><a href="Number.vue.html">Number.vue</a></td>
	<td data-value="71.43" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 71%;"></div><div class="cover-empty" style="width:29%;"></div></div></td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="3" class="abs low">0/3</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	</tr>

<tr>
	<td class="file medium" data-value="Password.vue"><a href="Password.vue.html">Password.vue</a></td>
	<td data-value="71.43" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 71%;"></div><div class="cover-empty" style="width:29%;"></div></div></td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="3" class="abs low">0/3</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	</tr>

<tr>
	<td class="file high" data-value="Radio.vue"><a href="Radio.vue.html">Radio.vue</a></td>
	<td data-value="80" class="pic high"><div class="chart"><div class="cover-fill" style="width: 80%;"></div><div class="cover-empty" style="width:20%;"></div></div></td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="8" class="abs medium">6/8</td>
	</tr>

<tr>
	<td class="file high" data-value="Select.vue"><a href="Select.vue.html">Select.vue</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	</tr>

<tr>
	<td class="file medium" data-value="Switch.vue"><a href="Switch.vue.html">Switch.vue</a></td>
	<td data-value="66.67" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 66%;"></div><div class="cover-empty" style="width:34%;"></div></div></td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="9" class="abs medium">6/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="3" class="abs low">1/3</td>
	<td data-value="62.5" class="pct medium">62.5%</td>
	<td data-value="8" class="abs medium">5/8</td>
	</tr>

<tr>
	<td class="file high" data-value="Textarea.vue"><a href="Textarea.vue.html">Textarea.vue</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri Jul 26 2019 11:02:13 GMT+0800 (GMT+08:00)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
