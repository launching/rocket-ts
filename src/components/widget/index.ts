import RInput from '@/components/widget/Input.vue';
import RSelect from '@/components/widget/Select.vue';
import RCheckbox from '@/components/widget/Checkbox.vue';
import RRadio from '@/components/widget/Radio.vue';
import RSwitch from '@/components/widget/Switch.vue';
import RTextarea from '@/components/widget/Textarea.vue';
import RPassword from '@/components/widget/Password.vue';
import RNumber from '@/components/widget/Number.vue';
import RDatePicker from '@/components/widget/DatePicker.vue';
import RDatePickerGroup from '@/components/widget/DatePickerGroup.vue';
import RRate from '@/components/widget/Rate.vue';

export default {
  RInput,
  RSelect,
  RCheckbox,
  RRadio,
  RSwitch,
  RTextarea,
  RPassword,
  RNumber,
  RDatePicker,
  RDatePickerGroup,
  RRate
};