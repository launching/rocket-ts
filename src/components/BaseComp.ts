import { Component, Prop, Vue, PropSync, Inject, Watch, Model } from 'vue-property-decorator';
import _ from 'lodash';
import { FormModel, FormItem, Validate, WidgetStore, ButtonItem } from '@/components/Interface';
@Component({})
export default class Comp extends Vue {
  @Prop()
  @Model('update:value')
  value?: any;

  targetOptions = [];
  @Prop(Boolean)
  disabled?: boolean;

  @Prop()
  options!: [{ value: string; label: string }];

  @Watch('store.value')
  onStoreValueChanged(val: any, oldVal: any) {
    this.$emit('update:value', val);
  }

  @Watch('value', { deep: true })
  refreshValue(val: any) {
    this.$set(this.store, 'value', val);
  }

  // 所有的信息都转异步
  async optionsHandler() {
    return _.isFunction(this.options) ? this.options() : this.options;
  }

  @Watch('option', { immediate: true, deep: true })
  async refreshOption() {
    const res: any = await this.optionsHandler();
    if (!Array(res)) {
      throw new Error('option is must be Array Type');
    }
    this.targetOptions = res.map((item: any) => {
      if (_.isString(item)) {
        return { label: item, value: 'item' };
      }

      return item;
    });
  }

  store: { value: any } = {
    value: this.value,
  };

  setValue(value: any) {
    this.store.value = value;
  }
}
