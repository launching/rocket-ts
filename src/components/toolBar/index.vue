<template lang="pug">
.r-toolbar
  el-button(v-for="child in children" :key="child.text" v-if="premiseStatus[child.text]"
  :type="child.type" :size="child.size" :icon="child.icon"
  @click="handlerClick(child)") {{child.text}}
</template>
<script lang="ts">
import { Component, Prop, Vue, Emit, Watch } from 'vue-property-decorator';
import _ from 'lodash';
import { ButtonItem } from '@/components/Interface';
@Component({})
export default class Toolbar extends Vue {
  @Prop(Object)
  readonly ctx?: object;

  @Prop({
    type: Array,
    default() {
      return [];
    },
  })
  readonly children!: Array<ButtonItem>;

  @Emit()
  async handlerClick(child: ButtonItem) {
    if (_.isFunction(child.action)) {
      await child.action(this.ctx);
    }
    return child;
  }

  premiseStatus: any = {};

  async premiseHandler(child: ButtonItem) {
    if (_.isFunction(child.premise)) {
      return child.premise(this.ctx);
    }
    return _.isUndefined(child.premise) ? true : child.premise;
  }

  @Watch('children', { immediate: true, deep: true })
  refreshPremise() {
    this.children.forEach(async (child) => {
      const status = await this.premiseHandler(child);
      this.$set(this.premiseStatus, child.text, status);
    });
  }
}
</script>
<style lang="stylus">
.r-toolbar
  display inline-block
</style>
