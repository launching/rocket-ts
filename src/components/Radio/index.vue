<template lang="pug">
el-radio-group(v-model="store.value")
  el-radio(v-for="option in targetOptions" :key="option.label" :label="option.label")
</template>
<script lang="ts">
import {
  Component,
  Prop,
  Vue,
  PropSync,
  Inject,
  Watch,
  Model,
  Emit,
  Mixins,
} from 'vue-property-decorator';
import { FormModel, FormItem } from '@/components/Interface';
import _ from 'lodash';
import BaseComp from '../BaseComp';

@Component({})
export default class RRadio extends Mixins(BaseComp) {}
</script>
<style lang="stylus" scoped></style>