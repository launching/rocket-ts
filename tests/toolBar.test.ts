import Vue from 'vue';
import ElementUI from 'element-ui';

import { shallowMount, mount } from '@vue/test-utils';
import RToolBar from '@/components/toolBar/index.vue';
import { ButtonItem } from '@/components/Interface';

Vue.use(ElementUI);

describe('ToolBar.vue', () => {
  const globalToolbar = [
    {
      type: 'primary',
      text: '立即创建',
      action: () => {},
    },
    {
      text: '新增域名',
      action: 'SUBMIT',
    },
    {
      text: '取消',
      action: 'CANCEL',
    },
  ];

  it('rend button', async () => {
    const wrapper = shallowMount(RToolBar, {
      propsData: {
        children: globalToolbar,
      },
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('click button targget click', async () => {
    const action = jest.fn();
    globalToolbar[0].action = action;
    const wrapper = mount(RToolBar, {
      propsData: {
        children: globalToolbar,
      },
    });

    await wrapper.vm.$nextTick();

    const button = wrapper.find('button');
    button.trigger('click');
    expect(action).toHaveBeenCalled();
  });

  it('premise is undefined button is show', async () => {
    const wrapper = mount(RToolBar, {
      propsData: {
        children: globalToolbar,
      },
    });
    await wrapper.vm.$nextTick();
    const buttons = wrapper.findAll('button');
    expect(buttons.length).toBe(3);
  });

  it('premise is true button is show', async () => {
    const toolbar = [
      {
        text: '新增域名',
        action: 'SUBMIT',
        premise: () => true,
      },
    ];
    const wrapper = mount(RToolBar, {
      propsData: {
        children: toolbar,
      },
    });
    await wrapper.vm.$nextTick();

    const buttons = wrapper.findAll('button');
    expect(buttons.length).toBe(1);
  });

  it('premise is true button is hidden', async () => {
    const toolbar = [
      {
        text: '新增域名',
        action: 'SUBMIT',
        premise() {
          return true;
        },
      },
      {
        text: '域名',
        action: 'SUBMIT',
        premise() {
          return false;
        },
      },
    ];

    const wrapper = mount(RToolBar, {
      propsData: {
        children: toolbar,
      },
    });
    await wrapper.vm.$nextTick();

    const buttons = wrapper.findAll('button');
    expect(buttons.length).toBe(1);
  });

  it('premise is promise hidden button is hidden', (callback) => {
    const toolbar = [
      {
        text: '新增域名',
        action: 'SUBMIT',
        premise() {
          return Promise.resolve(true);
        },
      },
      {
        text: '域名',
        action: 'SUBMIT',
        premise() {
          return Promise.resolve(false);
        },
      },
    ];
    const wrapper = mount(RToolBar, {
      propsData: {
        children: toolbar,
      },
    });

    setTimeout(() => {
      const buttons = wrapper.findAll('button');
      expect(buttons.length).toBe(1);
      callback();
    }, 100);
  });

  it('premise is Promise true after 1000s', (callback) => {
    const toolbar = [
      {
        text: '新增域名',
        action: 'SUBMIT',
        premise() {
          return new Promise((resolve) => {
            setTimeout(() => resolve(true), 1000);
          });
        },
      },
    ];
    const wrapper = mount(RToolBar, {
      propsData: {
        children: toolbar,
      },
    });

    setTimeout(() => {
      const buttons = wrapper.findAll('button');
      expect(buttons.length).toBe(1);
      callback();
    }, 1100);
  });
});
