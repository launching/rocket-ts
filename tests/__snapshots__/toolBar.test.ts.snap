// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolBar.vue rend button 1`] = `
<div class="r-toolbar">
  <el-button-stub type="primary" icon="" nativetype="button">立即创建</el-button-stub>
  <el-button-stub type="default" icon="" nativetype="button">新增域名</el-button-stub>
  <el-button-stub type="default" icon="" nativetype="button">取消</el-button-stub>
</div>
`;
